plugins/plugin-chart-echarts/src/Timeseries/transformProps.ts

```diff
@@ -0,0 +1,203 @@
+
+  const colorFn = CategoricalColorNamespace.getScale(colorScheme as string);
+
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-call
+  const rebasedData = rebaseTimeseriesDatum(queryData.data);
+  const rawSeries = extractTimeseriesSeries(rebasedData);
+
+  const series: echarts.EChartOption.Series[] = [];
+  const formatter = getNumberFormatter(contributionMode ? ',.0%' : yAxisFormat);
+
+  rawSeries.forEach(entry => {
+    const forecastSeries = extractForecastSeriesContext(entry.name || '');
+    const isConfidenceBand = [
+      ForecastSeriesEnum.ForecastLower,
+      ForecastSeriesEnum.ForecastUpper,
+    ].includes(forecastSeries.type);
+    const isObservation = forecastSeries.type === ForecastSeriesEnum.Observation;
+    const isTrend = forecastSeries.type === ForecastSeriesEnum.ForecastTrend;
+    let stackId;
+    if (isConfidenceBand) {
+      stackId = forecastSeries.name;
+    } else if (stack && isObservation) {
+      // the suffix of the observation series is '' (falsy), which disables
+      // stacking. Therefore we need to set something that is truthy.
+      stackId = 'obs';
+    } else if (stack && isTrend) {
+      stackId = forecastSeries.type;
+    }
+    let plotType;
+    if (!isConfidenceBand && (seriesType === 'scatter' || (forecastEnabled && isObservation))) {
+      plotType = 'scatter';
+    } else if (isConfidenceBand) {
+      plotType = 'line';
+    } else {
+      plotType = seriesType === 'bar' ? 'bar' : 'line';
+    }
+    const lineStyle = isConfidenceBand ? { opacity: 0 } : {};
+
+    if (!((stack || area) && isConfidenceBand))
+      series.push({
+        ...entry,
+        id: entry.name,
+        name: forecastSeries.name,
+        itemStyle: {
+          color: colorFn(forecastSeries.name),
+        },
+        type: plotType,
+        // @ts-ignore
+        smooth: seriesType === 'smooth',
+        step: ['start', 'middle', 'end'].includes(seriesType as string) ? seriesType : undefined,
+        stack: stackId,
+        lineStyle,
+        areaStyle: {
+          opacity: forecastSeries.type === ForecastSeriesEnum.ForecastUpper || area ? opacity : 0,
+        },
+        symbolSize:
+          !isConfidenceBand &&
+          (plotType === 'scatter' || (forecastEnabled && isObservation) || markerEnabled)
+            ? markerSize
+            : 0,
+      });
+  });
+  const echartOptions: echarts.EChartOption = {
+    grid: {
+      top: 30,
+      bottom: zoomable ? 80 : 0,
+      left: 20,
+      right: 20,
+      containLabel: true,
+    },
+    xAxis: { type: 'time' },
+    yAxis: {
+      type: logAxis ? 'log' : 'value',
+      min: contributionMode === 'row' && stack ? 0 : undefined,
+      max: contributionMode === 'row' && stack ? 1 : undefined,
+      minorTick: { show: true },
+      minorSplitLine: { show: minorSplitLine },
+      axisLabel: { formatter },
+    },
+    tooltip: {
+      trigger: 'axis',
+      formatter: (params, ticket, callback) => {
+        // @ts-ignore
```

Comment:
There's a handful of scattered `ts-ignore`s still... I'm not losing any sleep, but maybe we can take a pass at them and see how painful they are to strip out.

Reply:
These were actually quite painful, as the current `@types` definitions aren't always coherent. I'd defer the bulk of these unil ECharts 5.0 comes out, as it has way better typing built in to the library than the current 4.x branch.